src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -721,43 +716,76 @@ void ProcessVirtualMethod (MethodDefinition method)
-			if (!isInstantiated && !@base.IsAbstract && Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, method))
-				return;
+			return markForOverride;
+		}
 
-			// Only track instantiations if override removal is enabled and the type is instantiated.
-			// If it's disabled, all overrides are kept, so there's no instantiation site to blame.
-			if (Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, method) && isInstantiated) {
-				MarkMethod (method, new DependencyInfo (DependencyKind.OverrideOnInstantiatedType, method.DeclaringType), ScopeStack.CurrentScope.Origin);
+		/// <summary>
+		/// Marks the Override of <paramref name="overrideInformation"/> with the correct reason. Should be called when <see cref="ShouldMarkOverrideForBase(OverrideInformation, bool)"/> returns true.
+		/// </summary>
+		// TODO: Take into account a base method in preserved scope
+		void MarkOverrideForBaseMethod (OverrideInformation overrideInformation)
+		{
+			if (Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, overrideInformation.Override) && Annotations.IsInstantiated (overrideInformation.Override.DeclaringType)) {
+				MarkMethod (overrideInformation.Override, new DependencyInfo (DependencyKind.OverrideOnInstantiatedType, overrideInformation.Override.DeclaringType), ScopeStack.CurrentScope.Origin);
 			} else {
 				// If the optimization is disabled or it's an abstract type, we just mark it as a normal override.
-				Debug.Assert (!Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, method) || @base.IsAbstract);
-				MarkMethod (method, new DependencyInfo (DependencyKind.Override, @base), ScopeStack.CurrentScope.Origin);
+				Debug.Assert (!Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, overrideInformation.Override) || overrideInformation.Base.IsAbstract);
+				MarkMethod (overrideInformation.Override, new DependencyInfo (DependencyKind.Override, overrideInformation.Base), ScopeStack.CurrentScope.Origin);
 			}
+		}
 
-			if (method.IsVirtual)
-				ProcessVirtualMethod (method);
+		void MarkMethodIfNeededByBaseMethod (MethodDefinition method)
+		{
+			if (!Annotations.IsMarked (method.DeclaringType) || Annotations.GetBaseMethods (method) is not List<OverrideInformation> bases)
+				return;
+
+			foreach (var (ov, isDirectBase) in GetMarkedBaseMethods (bases)) {
+				if (ShouldMarkOverrideForBase (ov, isDirectBase))
+					MarkOverrideForBaseMethod (ov);
+			}
+
+			IEnumerable<(OverrideInformation Override, bool IsDirectBase)> GetMarkedBaseMethods (IEnumerable<OverrideInformation> overrides)
+			{
+				foreach (var ov in overrides) {
+					// Immediate base methods
+					if (Annotations.IsMarked (ov.Base) || IgnoreScope (ov.Base.DeclaringType.Scope))
+						yield return (ov, true);
+
+					// Bases of bases
+					else if (Annotations.GetBaseMethods (ov.Base) is IEnumerable<OverrideInformation> baseBases) {
```

Comment:
Is the `is` check redundant?

Reply:
The Annotations.Get[BaseMethods|Overrides|etc] can return null. I was thinking about changing the methods to return IEnumerable.Empty if not found in the cache, or adding an empty list to the dictionary, but thought there might have been a reason it was done this way before.

Reply:
You used type check not null check. I agree it should check for null but that could be simple `x == null` or `x is not null`